packages/core/src/actions.ts

```diff
@@ -505,3 +517,97 @@ export function escalate<
     }
   );
 }
+
+export function decide<TContext, TEvent extends EventObject>(
+  conds: DecideConditon<TContext, TEvent>[]
+): DecideAction<TContext, TEvent> {
+  return {
+    type: ActionTypes.Decide,
+    conds
+  };
+}
+
+export function resolveActions<TContext, TEvent extends EventObject>(
+  machine: StateNode<TContext, any, TEvent>,
+  currentState: State<TContext, TEvent> | undefined,
+  currentContext: TContext,
+  _event: SCXML.Event<TEvent>,
+  actions: ActionObject<TContext, TEvent>[]
+): [ActionObject<TContext, TEvent>[], TContext] {
+  const [assignActions, otherActions] = partition(
+    actions,
+    (action): action is AssignAction<TContext, TEvent> =>
+      action.type === actionTypes.assign
+  );
+
+  let updatedContext = assignActions.length
+    ? updateContext(currentContext, _event, assignActions, currentState)
+    : currentContext;
+
+  const resolvedActions = flatten(
+    otherActions.map(actionObject => {
+      switch (actionObject.type) {
+        case actionTypes.raise:
+          return resolveRaise(actionObject as RaiseAction<TEvent>);
+        case actionTypes.send:
+          const sendAction = resolveSend(
+            actionObject as SendAction<TContext, TEvent>,
+            updatedContext,
+            _event,
+            machine.options.delays
+          ) as ActionObject<TContext, TEvent>; // TODO: fix ActionTypes.Init
+
+          if (!IS_PRODUCTION) {
+            // warn after resolving as we can create better contextual message here
+            warn(
+              !isString(actionObject.delay) ||
+                typeof sendAction.delay === 'number',
+              // tslint:disable-next-line:max-line-length
+              `No delay reference for delay expression '${actionObject.delay}' was found on machine '${machine.id}'`
+            );
+          }
+
+          return sendAction;
+        case actionTypes.log:
+          return resolveLog(
+            actionObject as LogAction<TContext, TEvent>,
+            updatedContext,
+            _event
+          );
+        case actionTypes.decide:{
+          const decideAction = actionObject as DecideAction<TContext, TEvent>
+          const matchedActions = decideAction.conds.find(
+            condition => {
+              const guard = toGuard(condition.cond, machine.options.guards)
+              return !guard || evaluateGuard(machine, guard, updatedContext, _event, currentState as any)
+            }
+          )?.actions;
+
+          if (!matchedActions) {
+            return []
+          }
+
+          const resolved = resolveActions(machine, currentState, updatedContext, _event, toActionObjects(toArray(matchedActions)))
```

Comment:
this currently has a weird side effect of assigns outside of conditional actions being prioritized even if they come **after** conditional actions in document order, this will get fixed right away on the next branch thanks to the recent microstep work, but I would prefer to keep those quirky semantics here (maybe just with a note in the README?) as it simplifies the implementation

Reply:
Yep, that's fine